<a href="https://colab.research.google.com/github/prasadw350-lang/Assignments/blob/main/Assignment23.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [108]:
# Step 1: Import Libraries

import pandas as pd
import numpy as np
import joblib

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Step 2: Load Dataset
filepath = "/content/drive/MyDrive/insurance.csv"
df = pd.read_csv(filepath)

df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [109]:
# Step 3: Check Dataset

df.info()
df.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


In [110]:
df.isnull().sum()

,0
age,0
sex,0
bmi,0
children,0
smoker,0
region,0
charges,0


In [111]:
# Step 4: Convert Categorical Data into Numerical

df = pd.get_dummies(df, drop_first=True)
df.head()

,age,bmi,children,charges,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,16884.92400,False,True,False,False,True
1,18,33.770,1,1725.55230,True,False,False,True,False
2,28,33.000,3,4449.46200,True,False,False,True,False
3,33,22.705,0,21984.47061,True,False,True,False,False
4,32,28.880,0,3866.85520,True,False,True,False,False


In [112]:
# Step 5: Separate Features and Target

X = df.drop("charges", axis=1)
y = df["charges"]

In [113]:
# Step 6: Train-Test Split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [114]:
# Step 7: Feature Scaling

scaler = StandardScaler()

num_cols = ["age","bmi","children"]

X_train[num_cols] = scaler.fit_transform(X_train[num_cols])
X_test[num_cols] = scaler.transform(X_test[num_cols])

In [115]:
# Step 8: Train Linear Regression Model

model = LinearRegression()

model.fit(X_train, y_train)

LinearRegression()

In [116]:
# Step 9: Prediction

y_pred = model.predict(X_test)
print(y_pred[:10])

[ 8969.55027444  7068.74744287 36858.41091155  9454.67850053
 26973.17345656 10864.11316424   170.28084136 16903.45028662
  1092.43093614 11218.34318352]


In [117]:
# Step 10: Evaluate Model

r2 = r2_score(y_test, y_pred)
print("R2 Score:", r2)

R2 Score: 0.7835929767120722


In [118]:
# Step 11: Save Model

joblib.dump(model,"LR_Model.pkl")
joblib.dump(scaler,"LR_Scaler.pkl")
joblib.dump(X.columns.tolist(),"LR_Columns.pkl")

['LR_Columns.pkl']

## Logistic Regression (Mushroom Classification Dataset)

In [119]:
# Step 1: Import Libraries

import pandas as pd
import numpy as np
import joblib

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

# Step 2: Load Dataset
filepath = "/content/drive/MyDrive/mushrooms.csv"
df = pd.read_csv(filepath)

In [120]:
# Step 3: Check Dataset

df.info()
df.isnull().sum()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

,0
class,0
cap-shape,0
cap-surface,0
cap-color,0
bruises,0
odor,0
gill-attachment,0
gill-spacing,0
gill-size,0
gill-color,0


In [121]:
df.describe()

,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
count,8124,8124,8124,8124,8124,8124,8124,8124,8124,8124,...,8124,8124,8124,8124,8124,8124,8124,8124,8124,8124
unique,2,6,4,10,2,9,2,2,2,12,...,4,9,9,1,4,3,5,9,6,7
top,e,x,y,n,f,n,f,c,b,b,...,s,w,w,p,w,o,p,w,v,d
freq,4208,3656,3244,2284,4748,3528,7914,6812,5612,1728,...,4936,4464,4384,8124,7924,7488,3968,2388,4040,3148


In [122]:
# Step 4: Encode Categorical Columns

le = LabelEncoder()
for col in df.columns:
    df[col] = le.fit_transform(df[col])
df.head()

,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
0,1,5,2,4,1,6,1,0,1,4,...,2,7,7,0,2,1,4,2,3,5
1,0,5,2,9,1,0,1,0,0,4,...,2,7,7,0,2,1,4,3,2,1
2,0,0,2,8,1,3,1,0,0,5,...,2,7,7,0,2,1,4,3,2,3
3,1,5,3,8,1,6,1,0,1,5,...,2,7,7,0,2,1,4,2,3,5
4,0,5,2,3,0,5,1,1,0,4,...,2,7,7,0,2,1,0,3,0,1


In [123]:
# Step 5: Separate Features and Target

X = df.drop("class", axis=1)
y = df["class"]

# Step 6: Train-Test Split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [124]:
# Step 7: Feature Scaling

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Step 8: Train Logistic Regression Model

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [125]:
# Step 9: Prediction

y_pred = model.predict(X_test)
print(y_pred[:10])

[0 1 1 0 1 1 1 1 0 0]


In [126]:
# Step 10: Evaluation

cm = confusion_matrix(y_test, y_pred)
acc = accuracy_score(y_test, y_pred)
pre = precision_score(y_test, y_pred)
rec = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Confusion Matrix")
print(cm)
print("Accuracy :", acc)
print("Precision :", pre)
print("Recall :", rec)
print("F1 Score :", f1)

Confusion Matrix
[[799  44]
 [ 34 748]]
Accuracy : 0.952
Precision : 0.9444444444444444
Recall : 0.9565217391304348
F1 Score : 0.9504447268106735


In [127]:
# Step 11: Save Model

joblib.dump(model, "LogisticRegression_Model.pkl")
joblib.dump(scaler, "LogisticRegression_Scaler.pkl")
joblib.dump(X.columns.tolist(), "LogisticRegression_Columns.pkl")

['LogisticRegression_Columns.pkl']

In [128]:
logistic_acc = accuracy_score(y_test, y_pred)
logistic_pre = precision_score(y_test, y_pred)
logistic_rec = recall_score(y_test, y_pred)
logistic_f1 = f1_score(y_test, y_pred)

# KNN - (Mushroom Dataset)

In [129]:
# KNN Model

from sklearn.neighbors import KNeighborsClassifier
k_values = [3, 5, 7]
accuracy = []
for k in k_values:
    model = KNeighborsClassifier(n_neighbors=k)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    acc = accuracy_score(y_test, y_pred)
    accuracy.append(acc)

    print(f"K = {k}  Accuracy = {acc:.4f}")

K = 3  Accuracy = 1.0000
K = 5  Accuracy = 1.0000
K = 7  Accuracy = 1.0000


In [130]:
# Comparison Table

comparison = pd.DataFrame({
    "K Value": k_values,
    "Accuracy": accuracy
})

comparison

,K Value,Accuracy
0,3,1.0
1,5,1.0
2,7,1.0


In [131]:
# Best K

best_k = comparison.loc[comparison["Accuracy"].idxmax(), "K Value"]

print("Best K:", best_k)

Best K: 3


In [132]:
# Final Model

model = KNeighborsClassifier(n_neighbors=best_k)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

In [133]:
# Save Model

joblib.dump(model, "KNN_Model.pkl")

['KNN_Model.pkl']

In [134]:
knn_acc = accuracy_score(y_test, y_pred)
knn_pre = precision_score(y_test, y_pred)
knn_rec = recall_score(y_test, y_pred)
knn_f1 = f1_score(y_test, y_pred)

Naive Bayes

In [135]:
# Naive Bayes Model

from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [136]:
# Evaluation

cm = confusion_matrix(y_test, y_pred)
acc = accuracy_score(y_test, y_pred)
pre = precision_score(y_test, y_pred)
rec = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Confusion Matrix")
print(cm)
print("Accuracy :", acc)
print("Precision :", pre)
print("Recall :", rec)
print("F1 Score :", f1)

Confusion Matrix
[[771  72]
 [ 55 727]]
Accuracy : 0.9218461538461539
Precision : 0.9098873591989988
Recall : 0.9296675191815856
F1 Score : 0.9196710942441493


In [137]:
# Save Model

joblib.dump(model, "NaiveBayes_Model.pkl")

['NaiveBayes_Model.pkl']

In [138]:
naive_acc = accuracy_score(y_test, y_pred)
naive_pre = precision_score(y_test, y_pred)
naive_rec = recall_score(y_test, y_pred)
naive_f1 = f1_score(y_test, y_pred)

Same classification dataset, train and compare:

In [139]:
comparison = pd.DataFrame({
    "Algorithm": [
        "Logistic Regression",
        "K-Nearest Neighbors",
        "Naive Bayes"
    ],
    "Accuracy": [
        logistic_acc,
        knn_acc,
        naive_acc
    ],
    "Precision": [
        logistic_pre,
        knn_pre,
        naive_pre
    ],
    "Recall": [
        logistic_rec,
        knn_rec,
        naive_rec
    ],
    "F1 Score": [
        logistic_f1,
        knn_f1,
        naive_f1
    ]
})

comparison

,Algorithm,Accuracy,Precision,Recall,F1 Score
0,Logistic Regression,0.952000,0.944444,0.956522,0.950445
1,K-Nearest Neighbors,1.000000,1.000000,1.000000,1.000000
2,Naive Bayes,0.921846,0.909887,0.929668,0.919671


In [140]:
best_model = comparison.loc[comparison["Accuracy"].idxmax()]

print("Best Algorithm:", best_model["Algorithm"])
print("Accuracy:", best_model["Accuracy"])

Best Algorithm: K-Nearest Neighbors
Accuracy: 1.0


Write which algorithm performed best and why?

K-Nearest Neighbors (KNN) achieved the highest performance, with 100% Accuracy, Precision, Recall, and F1 Score on the mushroom dataset. Logistic Regression performed well with 95.20% Accuracy, while Naive Bayes achieved 92.18% Accuracy. Therefore, KNN is the best-performing algorithm because it correctly classified all test samples, resulting in perfect evaluation metrics.